backends/template/ceed-tmpl-sub.c

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2017-2018, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory. LLNL-CODE-734707.
+// All Rights reserved. See files LICENSE and NOTICE for details.
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+#include <ceed-backend.h>
+#include <string.h>
+
+static int CeedInit_Tmpl(const char *resource, Ceed ceed) {
+  int ierr;
+  if (strcmp(resource, "/cpu/self")
+      && strcmp(resource, "/cpu/self/tmpl/sub"))
+    return CeedError(ceed, 1, "Tmpl backend cannot use resource: %s", resource);
+
+  Ceed ceedref;
+  // Create refrence CEED that implementation will be dispatched
+  //   through unless overridden
+  CeedInit("/cpu/self/ref/blocked", &ceedref);
+  ierr = CeedSetDelegate(ceed, &ceedref); CeedChk(ierr);
+
+  // Create refrence CEED for objects
+  Ceed basisceedref;
+  CeedInit("/cpu/self/ref/blocked", &basisceedref);
+  ierr = CeedSetObjectDelegate(ceed, &basisceedref, "Basis");
+  CeedChk(ierr);
+  Ceed tensorceedref;
+  CeedInit("/cpu/self/ref/blocked", &tensorceedref);
+  ierr = CeedSetObjectDelegate(ceed, &tensorceedref, "TensorContract");
+  CeedChk(ierr);
+  Ceed opceedref;
+  CeedInit("/cpu/self/ref/blocked", &opceedref);
+  ierr = CeedSetObjectDelegate(ceed, &opceedref, "Operator");
+  CeedChk(ierr);
+
+  return 0;
+}
+
+__attribute__((constructor))
+static void Register(void) {
+  CeedRegister("/cpu/self/tmpl/sub", CeedInit_Tmpl, 70);
```

Comment:
Does this need to be a backend in the library or could it go in a test so we don't have to document and test it for all users?

Reply:
Perhaps we should replace the current `/cpu/self/tmpl` backend with `/cpu/self/tmpl/sub`?

This feature is intended to create backends from multiple different backends, so it feels like making a backend and running it through the test suite is the right way to test it?

I agree that the documentation needs to be improved.

As it is, I'm not wild about the template backend being built in a standard build of the library. Maybe we need a way to build the template backend(s) only for testing, like in Travis?

Reply:
We could build a `libceed-test.so` if all the tests need to run it, or we could just have tests that need it define it (by including a header or linking a common object file).

Reply:
I started the work but the Makefile dependencies are broken in parallel.

Reply:
Please have a look at https://github.com/CEED/libCEED/tree/jed/libceed-test

The `prove` target would still need to filter tmpl out of backends for non-t* (it's done in the makefile for `test` and `junit` targets.

Reply:
Branch is updated, but there is some Travis issue that I haven't been able to reproduce locally.

Reply:
I rebased to your branch and squashed out the debugging commits. I think we're doing pretty good right now. I left many of the commits, and 686de4a specifically as a separate so that I can more easily change our solution to the `--no-as-needed` issue in the future.